shell/packages/sandstorm-permissions/permissions.js

```diff
@@ -14,333 +14,773 @@
-// jscs:disable
-
-var Crypto = Npm.require("crypto");
-var Url = Npm.require("url");
+const Crypto = Npm.require("crypto");
+const Url = Npm.require("url");
 
 SandstormPermissions = {};
 
-function PermissionSet(array) {
+class PermissionSet {
   // A wrapper around an array of booleans.
 
-  if (!array) {
-    this.array = [];
-  } else if (array instanceof Array) {
-    this.array = array.slice(0);
-  } else {
-    throw new Error("don't know how to interpret as PermissionSet: " + array);
+  constructor(array) {
+    if (!array) {
+      this.array = [];
+    } else if (array instanceof Array) {
+      this.array = array.slice(0);
+    } else {
+      throw new Error("don't know how to interpret as PermissionSet: " + array);
+    }
   }
-}
 
-// Methods for mutating a PermissionSet by combining it with another PermissionSet.
-// These return a boolean indicating whether the operation had any effect.
+  static fromRoleAssignment(roleAssignment, viewInfo) {
+    let result = new PermissionSet([]);
+
+    if (!roleAssignment || "none" in roleAssignment) {
+      if (viewInfo.roles) {
+        for (let ii = 0; ii < viewInfo.roles.length; ++ii) {
+          const roleDef = viewInfo.roles[ii];
+          if (roleDef.default) {
+            result = new PermissionSet(roleDef.permissions);
+            break;
+          }
+        }
+      }
+    } else if ("allAccess" in roleAssignment) {
+      let length = 0;
+      if (viewInfo.permissions) {
+        length = viewInfo.permissions.length;
+      }
 
-PermissionSet.prototype.add = function(other) {
-  var changed = false;
-  for (var ii = 0; ii < other.array.length; ++ii) {
-    var old = this.array[ii];
-    this.array[ii] = this.array[ii] || other.array[ii];
-    if (old != this.array[ii]) {
-      changed = true;
+      const array = new Array(length);
+      for (let ii = 0; ii < array.length; ++ii) {
+        array[ii] = true;
+      }
+
+      result = new PermissionSet(array);
+    } else if ("roleId" in roleAssignment && viewInfo.roles && viewInfo.roles.length > 0) {
+      const roleDef = viewInfo.roles[roleAssignment.roleId];
+      if (roleDef) {
+        result = new PermissionSet(roleDef.permissions);
+      }
     }
+
+    if (roleAssignment) {
+      result.add(new PermissionSet(roleAssignment.addPermissionSet));
+      result.remove(new PermissionSet(roleAssignment.removePermissionSet));
+    }
+
+    return result;
   }
-  return changed;
-}
 
-PermissionSet.prototype.remove = function(other) {
-  var changed = false;
-  for (var ii = 0; ii < other.array.length && ii < this.array.length; ++ii) {
-    var old = this.array[ii];
-    this.array[ii] = this.array[ii] && !other.array[ii];
-    if (old != this.array[ii]) {
-      changed = true;
+  isSubsetOf(other) {
+    check(other, PermissionSet);
+    for (let ii = 0; ii < this.array.length; ++ii) {
+      const mine = this.array[ii];
+      const yours = other.array[ii] || false;
+      if (mine != yours) {
```

Comment:
Shouldn't e.g. [false, true] be considered a "subset" of [true, true]? It seems like the logic here is actually verifying that this PermissionSet is a prefix of `other` rather than a "subset". It would help to add some doc comments to the method explaining the expected behavior.